metrics-operator/controllers/common/providers/elk/elk.go

```diff
@@ -0,0 +1,121 @@
+package elasticsearch
+
+import (
+    "context"
+    "encoding/json"
+    "fmt"
+    "strings"
+    "time"
+
+    "github.com/go-logr/logr"
+    metricsapi "github.com/keptn/lifecycle-toolkit/metrics-operator/api/v1"
+    elastic "github.com/elastic/go-elasticsearch/v8"
+    "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+    warningLogStringElastic = "%s API returned warnings: %s"
+)
+
+type KeptnElasticProvider struct {
+    Log       logr.Logger
+    K8sClient client.Client
+    Elastic   *elastic.Client
+}
+
+func NewElasticProvider(log logr.Logger, k8sClient client.Client, elasticURL string) (*KeptnElasticProvider, error) {
+    cfg := elastic.Config{
+        Addresses: []string{
+            elasticURL,
+        },
+    }
+    es, err := elastic.NewClient(cfg)
+    if err != nil {
+        return nil, fmt.Errorf("failed to create elasticsearch client: %w", err)
+    }
+
+    return &KeptnElasticProvider{
+        Log:       log,
+        K8sClient: k8sClient,
+        Elastic:   es,
+    }, nil
+}
+
+func (r *KeptnElasticProvider) FetchAnalysisValue(ctx context.Context, query string, analysis metricsapi.Analysis, provider *metricsapi.KeptnMetricsProvider) (string, error) {
+    ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+    defer cancel()
+
+    result, err := r.runElasticQuery(ctx, query, analysis.GetFrom(), analysis.GetTo())
+    if err != nil {
+        return "", err
+    }
+
+    r.Log.Info(fmt.Sprintf("Elasticsearch query result: %v", result))
+    return r.extractMetric(result)
+}
+
+func (r *KeptnElasticProvider) EvaluateQuery(ctx context.Context, metric metricsapi.KeptnMetric, provider metricsapi.KeptnMetricsProvider) (string, []byte, error) {
+    ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+    defer cancel()
+
+    result, err := r.runElasticQuery(ctx, metric.Spec.Query, time.Now().Add(-30*time.Minute), time.Now())
```

Comment:
do we always want to have exactly the last 30min time slot?

Reply:
for now it's hardcoded but we can make it passed dynamically btw looking for your feedback on the right way?

Reply:
let's make it dynamic please